File: lang/parser/parse.ts
const keywords = {
    DECLARE_VAR_KEYWD: "let",
    DECLARE_CONST_KEYWD: "const"
}

const symbols = {
    SINGLE_LINE_COMMENT: "//"
}

const SyntaxTree = {
    program: {
        nodes: [] as {}[]
    }
};

function Parse(line: string){
    let tokens: string[] = line.split(" ");

    if(line.includes(symbols.SINGLE_LINE_COMMENT)){
        return;
    }

    if(tokens[0] == keywords.DECLARE_VAR_KEYWD){
        const variable = {
            type: "",
            identifier: "",
            initialValue: "" as string | null
        }

        let isTypeAnnotated = () => {
            if(tokens[1].includes(":")){
                let tk = tokens[1].split(":");
                console.log(tk)
                variable.identifier = tk[0];
                variable.type = typeof tk[1];

                return true;
            } else {
                variable.identifier = tokens[1];
                return false;
            }
        }

        if(tokens[2] == "="){
            variable.initialValue = tokens[3];
        } else {
            variable.initialValue = null;
        }

        if(!isTypeAnnotated() && variable.initialValue){
            variable.type = typeof variable.initialValue;
        }

        if(variable.type !== typeof variable.initialValue){
            console.log("MISMATCHING TYPE");
        }

        SyntaxTree.program.nodes.push({
            kind: "VAR_DECL",
            identifier: variable.identifier,
            value: variable.initialValue,
            type: variable.type
        });
    }
}

const code = [
    "let name:int = -123",
    "//this is a comment"
]

for(let i = 0; i < code.length; i++){
    Parse(code[i]);
}

console.log(SyntaxTree)

File: lang/parser/parse.cpp
/*
                    GNU GENERAL PUBLIC LICENSE
                       Version 3, 29 June 2007

 Copyright (C) 2007 Free Software Foundation, Inc. <https://fsf.org/>
 Everyone is permitted to copy and distribute verbatim copies
 of this license document, but changing it is not allowed.
*/

// Piece of code written by AndrewNation

#include <iostream>
#include <map>
#include <bits/stdc++.h>

std::map <std::string, std::string> keywords = {
    {"DECLARE_VAR_KEYWD", "let"},
    {"DECLARE_CONST_KEYWD", "const"}
};

std::map <std::string, std::string> symbols = {
    {"SINGLE_LINE_COMMENT", "//"}
};

void Parse(std::string line){
    std::stringstream tokens(line);
    std::string _tok;

    if(line.find(symbols["SINGLE_LINE_COMMENT"])){
        std::cout << "it is a comment";
        return;
    }
    
    while(tokens >> _tok){
        if(_tok == keywords["DECLARE_VAR_KEYWD"]){
        
        }
    }
}

int main() {
    const std::string testVarDecl = "let name:int = -123";
    const std::string comment = "//comment";
    
    Parse(testVarDecl);
    Parse(comment);
    return 0;
}
